Handle images with no user properly

msabramo opened the pull request:
This makes it so that images that don't have a user (e.g.: `ubuntu` rather than `gliderlabs/registrator`) work properly.

There have been several PRs I think that tried to tackle this (e.g.:  #100 , #104 & #97), but this one has tests (see https://github.com/kwk/docker-registry-frontend/pull/147/commits/9cb8ea158810371b057f43e324eb59536cd50456)! 😄 

Note that this changes the routing in a fairly significant way, because it switches to using query string variables for things like `tagsPerPage` and `tagPage`. The reason is that this helps to make the routes less ambiguous. For example, with the old routing, if you had a URI like `/repository/ubuntu/10` (the intent is that this means no user, image name is `ubuntu`, 10 tags per page), the routing could easily get confused and match a route like `/repository/:repositoryUser/:repositoryName`  and think that `repositoryUser` is `ubuntu` and `repositoryName` is `10`. Switching to query string variables removes this ambiguity, because `?tagsPerPage=10` makes it clear that the `10` is the number of tags per page and not the repository name.

Fixes #121 
Closes #97, #100, #104

msabramo committed:
Uncomment test for no repositoryUser

in `app/app.spec.js`, so I can work on getting this case working.

msabramo committed:
Add $log to repository-detail-controller

msabramo committed:
Set $scope.repository properly even if no user

Check if `$scope.repositoryUser` is defined and if it's not, set
`$scope.repository` simply to `$scope.repositoryName`.

msabramo committed:
Add route for repo with no repositoryUser

msabramo committed:
Make tagsPerPage dropdown use query string vars

instead of putting `tagsPerPage` in the route part of the URL, which
makes it difficult to distinguish between `repositoryName` and
`tagsPerPage`, because the routes are too similar.

msabramo committed:
Make first/next links use query string vars

instead of putting `tagsPerPage` in the route part of the URL, which
makes it difficult to distinguish between `repositoryName` and
`tagsPerPage`, because the routes are too similar.

msabramo committed:
Move defaultTagsPerPage

from `RepositoryListController` to `TagController`.

I think it makes more sense to read `defaultTagsPerPage` in
`TagController` than to do it in `RepositoryListController` and have to
pass it in URLs.

msabramo committed:
Fix "Details for repository" for repo w/ no user

msabramo committed:
Fix breadcrumb when no repositoryUser

kwk commented on app/repository/repository-detail-controller.js:
```diff
@@ -27,13 +34,13 @@ angular.module('repository-detail-controller', ['registry-services', 'app-mode-s
       if($scope.maxTagsPage > $scope.tagsCurrentPage){
         var nextPageNumber = $scope.tagsCurrentPage + 1;
-        return '/repository/'+$scope.repository+'/'+ $scope.tagsPerPage +'/' +nextPageNumber;
+        return '/repository/'+$scope.repository+'?tagsPerPage='+ $scope.tagsPerPage +'&tagPage=' +nextPageNumber;
```
I'm not sure that `tagsPerPage` can be a parameter. In line 63 of `app/app.js` this route is defined as 

``` javascript
when('/repository/:repositoryUser/:repositoryName/:tagsPerPage?/:tagPage?', {
```

As you can see, here it is part of the path and not a parameter.

Correct me, if I'm mistaken.

msabramo replied:
From https://www.sitepoint.com/premium/books/angularjs-novice-to-ninja/preview/using-routeparams-in-the-controller-6e0d5e2:

> Apart from named parameters, `$routeParams` also parses the query string, if present. So, if your URL contains a query string like `?key1=value1`, the key/value pair is added to `$routeParams`.

Also see the example at the bottom of https://docs.angularjs.org/api/ngRoute/service/$routeParams:

``` javascript
// Given:
// URL: http://server.com/index.html#/Chapter/1/Section/2?search=moby
// Route: /Chapter/:chapterId/Section/:sectionId
//
// Then
$routeParams ==> {chapterId:'1', sectionId:'2', search:'moby'}
```

Also I think you'll find it works if you checkout my branch and run it. Especially if you use Chrome and install the [ng-inspector for AngularJS Chrome extension](https://chrome.google.com/webstore/detail/ng-inspector-for-angularj/aadgmnobpdmgmigaicncghmmoeflnamj); with that extension installed, you can go to any page rendered by Angular and click the little "A" button in the toolbar and get a pane that allows you to browse the values of all the variables in `$scope`.

<img width="1437" alt="screen shot 2016-07-20 at 7 08 23 am" src="https://cloud.githubusercontent.com/assets/305268/16989370/d45424e8-4e48-11e6-8d63-0705d27e2ded.png">

I did forget to remove `:tagsPerPage?` from the routes though. I can update that.

msabramo replied:
OK, I:
- removed `:tagsPerPage` from the repository detail routes in 8e635d2421e0c089d7c20bc7c0c881412814c0ba.
- updated the unit tests to pass `tagsPerPage` as a query string parameter rather than in the route in 15d5cdb.

msabramo replied:
@kwk: Have I convinced you that `tagsPerPage` can be a query string parameter? If so, I think it also makes sense to do the same with `reposPerPage`...

kwk commented on app/repository/repository-detail-controller.js:
```diff
@@ -27,13 +34,13 @@ angular.module('repository-detail-controller', ['registry-services', 'app-mode-s
     $scope.getFirstHref = function (){
       if($scope.tagsCurrentPage > 1){
-        return '/repository/'+$scope.repository+'/' + $scope.tagsPerPage +'/1';
+        return '/repository/'+$scope.repository+'?tagsPerPage=' + $scope.tagsPerPage;
```
Same here.

msabramo replied:
Yeah, see https://github.com/kwk/docker-registry-frontend/pull/147#discussion_r71525785

kwk commented on app/repository/repository-detail.html:
```diff
@@ -40,14 +39,14 @@
         </a>
         <ul class="dropdown-menu">
-          <li><a href="repository/{{repository}}">Show all</a></li>
+          <li><a href="repository/{{repository}}?tagsPerPage=all">Show all</a></li>
```
Can `tagsPerPage` really be a parameter here and for the next few lines?

msabramo replied:
Yeah, see https://github.com/kwk/docker-registry-frontend/pull/147#discussion_r71525785

msabramo committed:
Remove tagsPerPage/tagPage from repo detail routes

msabramo committed:
Update unit tests to use tagsPerPage qs param

Updated the unit tests to page `tagsPerPage` as a query string parameter
rather than in the route.

msabramo committed:
Uncomment assertions about scope.reposPerPage

an make them work by casting string to an int with `parseInt`.

msabramo committed:
Make /tag work for image with no repositoryUser

msabramo committed:
repository-detail.html: Fix breadcrumb link

for case when there is no `repositoryUser`.

msabramo committed:
tag-detail.html: Fix breadcrumb link

when image has no `repositoryUser`.

msabramo committed:
tag-detail.html: Fix shown docker pull command

when image has no `repositoryUser`.

msabramo committed:
tag-detail.html: Show docker pull cmd in code font

I think it looks better.

msabramo commented:
@kwk: How does this look?

msabramo commented:
Awesome!
